crates/sui-core/src/authority.rs

```diff
@@ -3128,8 +3140,10 @@ impl AuthorityState {
                 epoch_store.get_capabilities(),
             );
 
+        // TODO: is the Move binary format version obtained from epoch store is the right one to be
+        // used for deserialization of next epoch's modules?
         let Some(next_epoch_system_package_bytes) = self.get_system_package_bytes(
-            next_epoch_system_packages.clone()
+            next_epoch_system_packages.clone(), epoch_store.protocol_config().move_binary_format_version()
```

Comment:
I think this is a great question, cc @mystenmark

Reply:
I think we have to use the current epoch's version. The package is being written during the current epoch, so it should abide by the rules of the current epoch.

Additionally, even if we wanted to use the next epoch, what would that mean in the case where there we do a protocol upgrade without a framework upgrade? Any framework may eventually be used in an epoch with a different protocol config, right?

Reply:
I have left this code as is then and updated the comment.

Reply:
> Additionally, even if we wanted to use the next epoch, what would that mean in the case where there we do a protocol upgrade without a framework upgrade?

I don't think this is an issue?  If we aren't doing a framework upgrade, we don't have to ask the question of which max version to serialize the framework at.

> Any framework may eventually be used in an epoch with a different protocol config, right?

Yes, that's right, there isn't a requirement that the protocol config exactly matches.  The precise scenario I'm thinking of is that when we run the protocol upgrade that introduces a new upperbound on the bytecode version, if we require that the framework is written at the old version, we will need to update our build process for the sui framework to serialize at the lower bytecode version initially, and then once that protocol upgrade has gone through, we can bump the bytecode version we build the sui-framework at back up.  I.e. it introduces an awkward two step process (similar to the current process of updating `SuiSystemState`) which is not the end of the world, but would be nice to avoid.

> The package is being written during the current epoch, so it should abide by the rules of the current epoch.

This is a valid argument, but there's also a complementary one that the package we're writing will only ever be read in the next epoch, so if we can infer from the fact that we're executing the upgrade that the binary understands the next epoch's bytecode format version, then we should be able to write the package out with the higher bound, but that's the condition I'm not clear is true (i.e. what happens to a node that's running a change epoch that does a protocol upgrade to a version that that node doesn't support).